filebeat/input/file/state.go

```diff
@@ -56,83 +56,85 @@ func (s *State) IsEmpty() bool {
 
 // States handles list of FileState
 type States struct {
-	states []State
+	states map[string]State
 	sync.RWMutex
 }
 
 func NewStates() *States {
 	return &States{
-		states: []State{},
+		states: map[string]State{},
 	}
 }
 
 // Update updates a state. If previous state didn't exist, new one is created
 func (s *States) Update(newState State) {
+	s.UpdateWithTimestamp(newState, time.Now())
+}
+
+// UpdateWithTimestamp updates a state, using the passed timestamp. If the
+// previous state didn't exist, a new entry is created.
+func (s *States) UpdateWithTimestamp(newState State, ts time.Time) {
 	s.Lock()
 	defer s.Unlock()
 
-	index, _ := s.findPrevious(newState)
-	newState.Timestamp = time.Now()
+	// ensure ID is set, so it won't be generated in find and on insert
+	id := newState.ID()
 
-	if index >= 0 {
-		s.states[index] = newState
-	} else {
-		// No existing state found, add new one
-		s.states = append(s.states, newState)
-		logp.Debug("input", "New state added for %s", newState.Source)
+	if logp.IsDebug("input") {
```

Comment:
Should we move this into `FindPrevious` and change the log message a bit? Seems to be out of place here a bit.

Reply:
No, as findPrevious is about searching. This message is about adding a new entry.